File: frontend/src/types/graph.ts
export type ClaimType = 'FACT' | 'ASSUMPTION' | 'PREDICTION' | 'OPINION'
export type LogicGate = 'or' | 'and'
export type CausalType = 'direct' | 'indirect' | 'probabilistic' | 'enabling' | 'inhibiting' | 'triggering'
export type ConditionType = 'sufficient' | 'necessary' | 'contributing'
export type BiasSeverity = 'low' | 'medium' | 'high'

export type { ReviewStatus } from './reasoning.ts'
import type { ReviewStatus } from './reasoning.ts'

export interface BiasWarning {
  type: string
  explanation: string
  severity: BiasSeverity
}

/** How a claim relates to the decision, in theory-based-view terms. */
export type DecisionRole = 'lever' | 'contingency' | 'mechanism' | 'outcome' | 'background'

export interface AnchorOption {
  key: string
  label: string
}

export interface AnchorOutcome {
  key: string
  label: string
  measure: string
}

/** The decision, stated so the pipeline can steer by it. */
export interface DecisionAnchor {
  decision: string
  options: AnchorOption[]
  outcomes: AnchorOutcome[]
  deadline: string
  constraints: string[]
  status: 'draft' | 'confirmed'
}

export interface CausalNode {
  id: string
  text: string
  claimType: ClaimType
  confidence: number
  belief: number | null
  sensitivity: number | null
  isCriticalPath: boolean
  isConvergencePoint?: boolean
  logicGate: LogicGate
  orderIndex: number
  sourceSentence: string | null
  beliefLow: number | null
  beliefHigh: number | null
  // Human review state
  reviewStatus: ReviewStatus
  isActive: boolean
  userNote: string | null
  /** How firmly the SOURCE asserts this. A property of the text, not the world. */
  confidence_isAssertion?: never
  /** Probability the claim is actually true. What propagation is seeded with. */
  prior: number
  /** Source sentences of duplicate claims merged into this one. */
  corroboratedBy: string[] | null
  duplicateCount: number
  // Relation to the decision anchor. Optional: absent on unanchored projects
  // and on nodes streamed mid-run.
  /** 'ai', 'user', or 'frame' for an outcome node created from the anchor. */
  origin?: string
  decisionRole?: DecisionRole | null
  /** The model's 0-1 reading of how much this claim could change the choice. */
  relevance?: number | null
  relevanceReason?: string | null
  /** Option and outcome keys (O1, Y2 ...) this claim bears on. */
  bearsOn?: string[] | null
  /** Causal hops to the nearest outcome; null when no path exists. */
  anchorDistance?: number | null
  effectiveRelevance?: number | null
  /** Read as background by the model, yet on a causal path to an outcome. */
  isPeripheral?: boolean
  /** Shown by the decision lens. */
  inLens?: boolean
  // Layout computed
  x?: number
  y?: number
  depth?: number
  collapsed?: boolean
}

export interface CausalEdge {
  id: string
  sourceId: string
  targetId: string
  mechanism: string
  strength: number
  timeDelay: string | null
  conditions: string[] | null
  reversible: boolean
  evidenceScore: number
  causalType: CausalType
  conditionType: ConditionType
  temporalWindow: string | null
  decayType: string
  biasWarnings: BiasWarning[]
  consensusLevel: string
  sensitivity: number | null
  isFeedback: boolean
  evidences: Evidence[]
  // Human review state
  reviewStatus: ReviewStatus
  isActive: boolean
  userNote: string | null
  strengthOverride: number | null
  /** strengthOverride when set, otherwise strength. */
  effectiveStrength: number
  /** Size of the effect if the link is real. */
  effect: number
  /** Certainty the link exists at all. `strength` is effect x this. */
  linkConfidence: number
  /** What the proposing call scored, before blind re-scoring. */
  authorEffect: number | null
  authorConfidence: number | null
  blindScored: boolean
  /** author weight minus blind weight; positive means the author over-scored. */
  inflation: number | null
}

export interface Evidence {
  id: string
  evidenceType: 'supporting' | 'contradicting'
  sourceUrl: string
  sourceTitle: string
  sourceType: string
  snippet: string
  relevanceScore: number
  credibilityScore: number
  sourceTier: number
}


export interface CausalGraph {
  projectId: string
  nodes: CausalNode[]
  edges: CausalEdge[]
  criticalPath: string[]
  hasTemporal: boolean
  graphRevision: number
  decisionObjective: string | null
  decisionAnchor?: DecisionAnchor | null
}

export interface BeliefChange {
  oldBelief: number
  newBelief: number
  delta: number
}

export interface PathInfo {
  path: string[]
  compoundProbability: number
}

export interface TemporalBeliefSample {
  time: number    // days from origin
  belief: number  // 0..1
}

export type ViewMode = 'panorama' | 'focus' | 'compare' | 'timeline'

File: decision_studio/api/models/graph.py
"""Pydantic schemas for graph endpoints."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class ClaimResponse(BaseModel):
    """A single claim node in the causal graph."""

    id: UUID
    text: str
    claim_type: str
    confidence: float
    belief: float | None = None  # From belief propagation
    sensitivity: float | None = None
    is_critical_path: bool = False
    is_convergence_point: bool = False
    logic_gate: str = "or"
    order_index: int
    source_sentence: str | None = None
    belief_low: float | None = None
    belief_high: float | None = None
    #: The belief if causes that share a driver move together (graph/shared_causes.py).
    #: Set only where it differs materially from ``belief``.
    belief_if_dependent: float | None = None
    #: The upstream claims those causes share. Empty when the gap is inherited
    #: from a claim further up.
    shared_causes: list[str] = []
    # --- Human review state (decision reasoning) ---
    review_status: str = "accepted"
    is_active: bool = True
    user_note: str | None = None
    # --- Assertion firmness vs truth probability (they are different) ---
    prior: float = 0.5
    corroborated_by: list[str] | None = None
    duplicate_count: int = 0
    # --- Relation to the decision anchor (null without an anchor) ---
    origin: str = "ai"
    decision_role: str | None = None
    relevance: float | None = None
    relevance_reason: str | None = None
    bears_on: list[str] | None = None
    #: Causal hops to the nearest outcome node; null with no path.
    anchor_distance: int | None = None
    #: max(stated relevance, structural relevance). See graph/anchoring.py.
    effective_relevance: float | None = None
    #: Read as background by the model, yet on a causal path to an outcome.
    is_peripheral: bool = False
    #: Shown by the decision lens by default.
    in_lens: bool = True


class EvidenceResponse(BaseModel):
    """Evidence supporting or contradicting a causal edge."""

    id: UUID
    evidence_type: str
    source_url: str
    source_title: str
    source_type: str
    snippet: str
    relevance_score: float
    credibility_score: float
    source_tier: int = 4
    freshness_score: float = 0.5
    published_date: datetime | None = None


class EdgeResponse(BaseModel):
    """A causal edge between two claims."""

    id: UUID
    source_claim_id: UUID
    target_claim_id: UUID
    mechanism: str
    strength: float
    time_delay: str | None = None
    conditions: list[str] | None = None
    reversible: bool
    evidence_score: float
    causal_type: str = "direct"
    condition_type: str = "contributing"
    temporal_window: str | None = None
    decay_type: str = "none"
    bias_warnings: list[dict] = []
    consensus_level: str = "insufficient"
    sensitivity: float | None = None
    is_feedback: bool = False
    evidences: list[EvidenceResponse] = []
    statistical_validation: str | None = None  # confirmed, unsupported, contradicted, not_tested
    stat_p_value: float | None = None
    stat_f_statistic: float | None = None
    stat_effect_size: float | None = None
    stat_lag: int | None = None
    # --- Human review state (decision reasoning) ---
    review_status: str = "accepted"
    is_active: bool = True
    user_note: str | None = None
    strength_override: float | None = None
    effective_strength: float | None = None
    # --- Effect vs link confidence (strength is their product) ---
    effect: float = 0.5
    link_confidence: float = 0.5
    author_effect: float | None = None
    author_confidence: float | None = None
    blind_scored: bool = False
    inflation: float | None = None


class GraphResponse(BaseModel):
    """Complete graph data for a project, including computed attributes."""

    project_id: UUID
    claims: list[ClaimResponse]
    edges: list[EdgeResponse]
    critical_path: list[UUID] = []
    has_temporal: bool = True
    graph_revision: int = 1
    decision_objective: str | None = None
    decision_anchor: dict | None = None


class EdgeUpdateRequest(BaseModel):
    """Request body for updating an edge's strength."""

    strength: float = Field(..., ge=0.0, le=1.0)


# --- Graph Operations ---


class ExpandRequest(BaseModel):
    """Request to expand consequences from a node."""

    node_id: UUID
    user_reasoning: str | None = Field(None, max_length=2000)


class TraceBackRequest(BaseModel):
    """Request to trace back causes to a node."""

    node_id: UUID
    user_reasoning: str | None = Field(None, max_length=2000)


class ChallengeRequest(BaseModel):
    """Request to challenge an edge with fresh evidence."""

    edge_id: UUID
    user_reasoning: str | None = Field(None, max_length=2000)


class WhatIfModification(BaseModel):
    """A single modification for what-if analysis."""

    type: str = Field(..., description="edge_strength or node_probability")
    target_id: UUID
    value: float = Field(..., ge=0.0, le=1.0)
    source_id: UUID | None = None


class WhatIfRequest(BaseModel):
    """Request for what-if analysis with modifications."""

    modifications: list[WhatIfModification]


class BeliefChange(BaseModel):
    """Belief change for a single node."""

    old_belief: float
    new_belief: float
    delta: float


class GraphOperationResult(BaseModel):
    """Result of expand or trace-back operations."""

    new_nodes: list[ClaimResponse]
    new_edges: list[EdgeResponse]
    converged_edges: list[EdgeResponse]
    graph: GraphResponse


class ChallengeResult(BaseModel):
    """Result of a challenge operation."""

    edge_id: UUID
    new_evidence_score: float
    new_evidences: list[EvidenceResponse]
    belief_changes: dict[str, BeliefChange]
    graph: GraphResponse


class WhatIfResult(BaseModel):
    """Result of what-if analysis (non-persistent)."""

    changes: dict[str, BeliefChange]
    modified_graph: GraphResponse


class PathInfo(BaseModel):
    """A single path to a node with compound probability."""

    path: list[UUID]
    compound_probability: float


class FocusResult(BaseModel):
    """Result of focus subgraph computation."""

    focus_node_id: UUID
    visible_node_ids: list[UUID]
    paths: list[PathInfo]


# --- Strategic Advisor ---


class PerspectiveSuggestion(BaseModel):
    """A single AI-generated perspective suggestion."""

    label: str
    description: str


class SuggestPerspectivesResult(BaseModel):
    """Result of perspective suggestion generation."""

    suggestions: list[PerspectiveSuggestion]


class AdviseRequest(BaseModel):
    """Request for strategic advisory analysis."""

    user_context: str = Field(..., min_length=10, max_length=5000)
    perspective_tags: list[str] = []
    session_id: str | None = None


class DirectImpact(BaseModel):
    claim_text: str
    impact_description: str
    severity: str
    timeline: str


class IndirectImpact(BaseModel):
    causal_chain: str
    impact_description: str
    severity: str
    timeline: str


class ImpactAssessment(BaseModel):
    summary: str
    overall_severity: str
    direct_impacts: list[DirectImpact]
    indirect_impacts: list[IndirectImpact]


class Prediction(BaseModel):
    prediction: str
    probability: float
    timeframe: str
    basis: str
    confidence_note: str


class RecommendedAction(BaseModel):
    action: str
    priority: str
    timeframe: str
    addresses_claim: str
    rationale: str


class EscalationScenario(BaseModel):
    scenario_name: str
    trigger: str
    causal_chain: str
    impact_on_user: str
    probability: float
    severity: str
    contingency_actions: list[str]


class KeyIndicator(BaseModel):
    indicator: str
    data_source: str
    related_claim: str
    threshold: str
    signal_type: str


class AdviseResult(BaseModel):
    """Result of strategic advisory analysis."""

    impact_assessment: ImpactAssessment
    predictions: list[Prediction]
    recommended_actions: list[RecommendedAction]
    escalation_scenarios: list[EscalationScenario]
    key_indicators: list[KeyIndicator]
